src/funcx_common/task_storage/s3.py

```diff
@@ -0,0 +1,50 @@
+import typing as t
+
+import boto3
+
+from ..tasks import TaskProtocol
+from .base import StorageException, TaskStorage
+
+
+class S3TaskStorage(TaskStorage):
+    """
+    Store Task data in S3
+
+    """
+
+    storage_id = "MemoryTaskStorage"
+
+    def __init__(self, bucket_name: str) -> None:
+        self.bucket_name = bucket_name
+        self.client = boto3.client("s3")
+        self._results: t.Dict[str, str] = {}
+
+    def store_result(
+        self, task: TaskProtocol, result: str, ACL: str = "authenticated"
+    ) -> bool:
+        key = f"{task.task_id}.result"
+        self.client.put_object(  # ACL=ACL,
+            Body=result.encode("utf-8"),
+            Bucket=self.bucket_name,
+            Key=key,
+        )
+        task.result_reference = {
+            "storage_id": self.storage_id,
+            "s3bucket": self.bucket_name,
+            "key": key,
+        }
+        return True
+
+    def get_result(self, task: TaskProtocol) -> t.Optional[str]:
+        if (
+            task.result_reference
+            and task.result_reference["storage_id"] == self.storage_id
+        ):
+            response = self.client.get_object(
+                Bucket=task.result_reference["s3bucket"],
+                Key=task.result_reference["key"],
+            )
```

Comment:
In the event that the object is not there, this will raise an error which we currently do not catch.
Wouldn't that be a case for the interface returning `None`?

Reply:
If we can't fetch, shouldn't it be a storage error?

Reply:
That would make this highly asymmetric with the Redis behavior, in which `task.result` is `None` if there is no data.

Raising an error (of an arbitrary type) on a failed get poses a problem if we want a future version of this to failover in the case that S3 data is missing.

Consider what the usage looks like downstream when someone wants to write code which attempts to fetch a result and handles the potential for a failure:
```python
task_storage = SomeComplexFuturisticTaskStorage(..., S3TaskStorage(...), ...)
...
result = task_storage.get_result(task)
if result is None:
    raise MissingResultError(...)
return result
```
vs
```python
try:
    return task_storage.get_result(task)
except botocore.exceptions.ClientError as err:
    if _is_s3_404(err):
        raise MissingResultException(...)
    raise
except otherlibrary.OtherMissingDataError:
    raise MissingResultException(...)
```

The interface defines this method as returning `Optional[str]` in order to make missing data part of the interface. If S3 missing data raises an error, then Redis missing data should also raise an error (rather than returning `None`). We can definitely go down that path, but I don't think it will make our lives easier when we're using this stuff.

Reply:
> That would make this highly asymmetric with the Redis behavior, in which task.result is None if there is no data.

I'm probably missing something, but looking at [lines 39-40](https://github.com/funcx-faas/funcx-common/pull/5/files#diff-f4e2f807ccfb5e475f899cbf4871304e8578d1307a5ebbd5e14dbd33b72a66f0R40), redis *does* raise a storage error if task.result is None.

Reply:
I'm not sure why you think it's asymmetric. Both RedisStorage and S3Storage will raise an exception if the result can't be fetched. If the concern is that S3Storage could raise boto3 client exceptions, I can add additional handling around that. It's always the result or an exception.

Currently, RedisStorage raises a StorageException if task.result is None, or if the storage_id is not matching in the task.result_reference.

Reply:
That definitely changed at some point; there was a version of this in which `RedisStorage` would return `task.result` as long as the `task.result_reference` lined up.

If this is always returning the result or an exception, why is it still `Optional[str]`? What is the case in which one of these methods could return `None`?